openapi_metadata.go

```diff
@@ -0,0 +1,344 @@
+// slice and registeredNames map.
+func NewMetadataParsers() *MetadataParsers {
+	return &MetadataParsers{
+		registeredParsers: []MetadataParserEntry{},
+		registeredNames:   make(map[string]bool),
+	}
+}
+
+// GetRegisteredParsers returns a slice of all currently registered metadata parsers.
+// This method ensures thread-safe access to the registeredParsers slice by using a mutex lock.
+func (mp *MetadataParsers) GetRegisteredParsers() []MetadataParserEntry {
+	mp.parserLock.Lock()
+	defer mp.parserLock.Unlock()
+
+	return mp.registeredParsers
+}
+
+// Reset clears all registered metadata parsers from the slice and resets the
+// registeredNames map as empty. This method ensures thread-safe access to the
+// registeredParsers slice and registeredNames map by using a mutex lock.
+func (mp *MetadataParsers) Reset() {
+	mp.parserLock.Lock()
+	defer mp.parserLock.Unlock()
+
+	mp.registeredParsers = []MetadataParserEntry{}
+	mp.registeredNames = make(map[string]bool)
+}
+
+// InitializeMetadataParsers initializes the metadata parsers with the given customParsers.
+// It will append the customParsers to the existing DefaultParsers. If a parser with the
+// same name is already registered, it will be skipped.
+func (mp *MetadataParsers) InitializeMetadataParsers(customParsers []MetadataParserEntry) {
+	mp.parserLock.Lock()
+	defer mp.parserLock.Unlock()
+
+	mp.registeredParsers = []MetadataParserEntry{}
+	mp.registeredNames = make(map[string]bool)
+
+	var parsersToRegister []MetadataParserEntry
+	parsersToRegister = customParsers
+
+	for _, entry := range parsersToRegister {
+		if _, exists := mp.registeredNames[entry.Name]; exists {
+			slog.Warn("Parser already registered", "name", entry.Name)
+			continue
+		}
+		mp.registeredNames[entry.Name] = true
+		mp.registeredParsers = append(mp.registeredParsers, entry)
+	}
+}
+
+// RegisterMetadataParser registers a new metadata parser with the given name, parser
+// function, and position string. The position string can be one of the following:
+// "start", "end", "before", or "after". If the position is "before" or "after", you
+// must also provide the name of the parser that you want to place the new parser
+// before or after. If the relative parser is not found, an error is returned. If the
+// parser is already registered, the function does nothing and returns nil. The
+// function is thread-safe and ensures that the registeredParsers slice and
+// registeredNames map are accessed in a thread-safe manner.
+func (mp *MetadataParsers) RegisterMetadataParser(name string, parser MetadataParserFunction, position string, relativeTo string) error {
+	validPositions := map[string]bool{"start": true, "end": true, "before": true, "after": true}
+	if !validPositions[position] {
+		return fmt.Errorf("Invalid position. Use 'start', 'end', 'before', or 'after'")
+	}
+	if (position == "before" || position == "after") && mp.findParserIndex(relativeTo) == -1 {
```

Comment:
Before, after, end, start should be constants prefixed with position or something